File: rlm/cli/tui/live_api.py
"""Cliente HTTP do workbench vivo para o servidor Arkhe."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any
from urllib import error as uerror
from urllib import request as urequest

from rlm.cli.context import CliContext


class LiveWorkbenchError(RuntimeError):
    """Erro operacional ao conversar com o backend vivo do workbench."""


def _internal_token(env: dict[str, str]) -> str:
    for name in ("RLM_INTERNAL_TOKEN", "RLM_WS_TOKEN", "RLM_API_TOKEN"):
        token = env.get(name, "").strip()
        if token:
            return token
    return ""


def _build_headers(env: dict[str, str]) -> dict[str, str]:
    headers = {"Content-Type": "application/json"}
    token = _internal_token(env)
    if token:
        headers["X-RLM-Token"] = token
    return headers


@dataclass(frozen=True, slots=True)
class LiveSessionInfo:
    session_id: str
    client_id: str
    status: str
    state_dir: str
    metadata: dict[str, Any] = field(default_factory=dict)


class LiveWorkbenchAPI:
    def __init__(self, context: CliContext) -> None:
        self._context = context
        self._base_url = context.env.get("RLM_INTERNAL_HOST", context.api_base_url()).rstrip("/")
        self._headers = _build_headers(dict(context.env))

    @property
    def base_url(self) -> str:
        return self._base_url

    def probe(self, *, timeout: int = 3) -> bool:
        """Verifica se o servidor vivo esta acessivel via /health."""
        url = f"{self._base_url}/health"
        req = urequest.Request(url, headers=self._headers, method="GET")
        try:
            with urequest.urlopen(req, timeout=timeout) as resp:
                return resp.status == 200
        except Exception:
            return False

    def ensure_session(self, client_id: str) -> LiveSessionInfo:
        payload = self._request_json("POST", "/operator/session", {"client_id": client_id})
        return LiveSessionInfo(
            session_id=str(payload.get("session_id", "")),
            client_id=str(payload.get("client_id", client_id)),
            status=str(payload.get("status", "idle")),
            state_dir=str(payload.get("state_dir", "")),
            metadata=dict(payload.get("metadata") or {}),
        )

    def fetch_activity(self, session_id: str) -> dict[str, Any]:
        return self._request_json("GET", f"/operator/session/{session_id}/activity")

    def dispatch_prompt(self, session_id: str, client_id: str, text: str) -> dict[str, Any]:
        return self._request_json(
            "POST",
            f"/operator/session/{session_id}/message",
            {"client_id": client_id, "text": text},
            timeout=15,
        )

    def apply_command(
        self,
        session_id: str,
        *,
        client_id: str,
        command_type: str,
        payload: dict[str, Any],
        branch_id: int | None,
    ) -> dict[str, Any]:
        return self._request_json(
            "POST",
            f"/operator/session/{session_id}/commands",
            {
                "client_id": client_id,
                "command_type": command_type,
                "payload": payload,
                "branch_id": branch_id,
            },
        )

    def _request_json(
        self,
        method: str,
        path: str,
        body: dict[str, Any] | None = None,
        *,
        timeout: int = 10,
    ) -> dict[str, Any]:
        url = f"{self._base_url}{path}"
        data = None
        if body is not None:
            data = json.dumps(body, ensure_ascii=False).encode("utf-8")
        req = urequest.Request(url, data=data, headers=self._headers, method=method)
        try:
            with urequest.urlopen(req, timeout=timeout) as resp:
                payload = json.loads(resp.read().decode("utf-8"))
        except uerror.HTTPError as exc:
            detail = exc.read().decode("utf-8", errors="replace")
            raise LiveWorkbenchError(f"HTTP {exc.code} em {path}: {detail[:300]}") from exc
        except uerror.URLError as exc:
            raise LiveWorkbenchError(
                f"Servidor vivo indisponivel em {self._base_url}. Rode 'arkhe start' e tente novamente."
            ) from exc
        except Exception as exc:
            raise LiveWorkbenchError(f"Falha ao chamar backend vivo {path}: {exc}") from exc

        if not isinstance(payload, dict):
            raise LiveWorkbenchError(f"Resposta invalida do backend vivo em {path}")
        return payload

    # ── Channel status endpoints ──────────────────────────────────────

    def fetch_channels_status(self) -> dict[str, Any]:
        """GET /api/channels/status — retorna snapshot de todos os canais."""
        return self._request_json("GET", "/api/channels/status")

    def probe_channel(self, channel_id: str) -> dict[str, Any]:
        """POST /api/channels/{channel_id}/probe — probe sob demanda."""
        return self._request_json("POST", f"/api/channels/{channel_id}/probe")

    def cross_channel_send(self, target_client_id: str, message: str) -> dict[str, Any]:
        """POST /api/channels/send — envia mensagem cross-channel."""
        return self._request_json(
            "POST",
            "/api/channels/send",
            {"target_client_id": target_client_id, "message": message},
        )

File: packages/cli/src/tui/live-api.ts
/**
 * Cliente HTTP do workbench vivo para o servidor Arkhe.
 *
 * Migrado de rlm/cli/tui/live_api.py
 */

import type { CliContext } from "../context.js";

// ─────────────────────────────────────────────── Erro operacional

export class LiveWorkbenchError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "LiveWorkbenchError";
  }
}

// ─────────────────────────────────────────────── Helpers internos

function getInternalToken(env: Record<string, string>): string {
  for (const name of ["RLM_INTERNAL_TOKEN", "RLM_WS_TOKEN", "RLM_API_TOKEN"]) {
    const token = (env[name] ?? "").trim();
    if (token) return token;
  }
  return "";
}

function buildHeaders(env: Record<string, string>): Record<string, string> {
  const headers: Record<string, string> = { "Content-Type": "application/json" };
  const token = getInternalToken(env);
  if (token) {
    headers["X-RLM-Token"] = token;
  }
  return headers;
}

// ─────────────────────────────────────────────── Tipos de dados

export interface LiveSessionInfo {
  session_id: string;
  client_id: string;
  status: string;
  state_dir: string;
  metadata: Record<string, unknown>;
}

// ─────────────────────────────────────────────── Classe principal

export class LiveWorkbenchAPI {
  private readonly _context: CliContext;
  private readonly _baseUrl: string;
  private readonly _headers: Record<string, string>;

  constructor(context: CliContext, baseUrl?: string) {
    this._context = context;
    this._baseUrl = (
      baseUrl ||
      (context.env as Record<string, string>)["RLM_INTERNAL_HOST"] ||
      context.apiBaseUrl()
    ).replace(/\/$/, "");
    this._headers = buildHeaders(context.env as Record<string, string>);
  }

  get baseUrl(): string {
    return this._baseUrl;
  }

  /** Verifica se o servidor está acessível via /health. */
  async probe(opts: { timeout?: number } = {}): Promise<boolean> {
    const url = `${this._baseUrl}/health`;
    try {
      const resp = await fetch(url, {
        method: "GET",
        headers: this._headers,
        signal: AbortSignal.timeout((opts.timeout ?? 3) * 1000),
      });
      return resp.status === 200;
    } catch {
      return false;
    }
  }

  /** Cria ou retoma uma sessão de operador. */
  async ensureSession(clientId: string): Promise<LiveSessionInfo> {
    const payload = await this._requestJson("POST", "/operator/session", { client_id: clientId });
    return {
      session_id: String(payload["session_id"] ?? ""),
      client_id: String(payload["client_id"] ?? clientId),
      status: String(payload["status"] ?? "idle"),
      state_dir: String(payload["state_dir"] ?? ""),
      metadata: (payload["metadata"] as Record<string, unknown>) ?? {},
    };
  }

  /** Busca atividade da sessão. */
  async fetchActivity(sessionId: string): Promise<Record<string, unknown>> {
    return this._requestJson("GET", `/operator/session/${sessionId}/activity`);
  }

  /** Envia um prompt para a sessão. */
  async dispatchPrompt(
    sessionId: string,
    clientId: string,
    text: string
  ): Promise<Record<string, unknown>> {
    return this._requestJson(
      "POST",
      `/operator/session/${sessionId}/message`,
      { client_id: clientId, text },
      { timeout: 15 }
    );
  }

  /** Aplica um comando à sessão. */
  async applyCommand(
    sessionId: string,
    opts: {
      clientId: string;
      commandType: string;
      payload: Record<string, unknown>;
      branchId: number | null;
    }
  ): Promise<Record<string, unknown>> {
    return this._requestJson("POST", `/operator/session/${sessionId}/commands`, {
      client_id: opts.clientId,
      command_type: opts.commandType,
      payload: opts.payload,
      branch_id: opts.branchId,
    });
  }

  // ── Endpoints de status de canais ──────────────────────────────────────

  /** GET /api/channels/status — retorna snapshot de todos os canais. */
  async fetchChannelsStatus(): Promise<Record<string, unknown>> {
    return this._requestJson("GET", "/api/channels/status");
  }

  /** POST /api/channels/{channelId}/probe — probe sob demanda. */
  async probeChannel(channelId: string): Promise<Record<string, unknown>> {
    return this._requestJson("POST", `/api/channels/${channelId}/probe`);
  }

  /** POST /api/channels/send — envia mensagem cross-channel. */
  async crossChannelSend(
    targetClientId: string,
    message: string
  ): Promise<Record<string, unknown>> {
    return this._requestJson("POST", "/api/channels/send", {
      target_client_id: targetClientId,
      message,
    });
  }

  // ── Método interno de requisição ───────────────────────────────────────

  private async _requestJson(
    method: string,
    apiPath: string,
    body?: Record<string, unknown>,
    opts: { timeout?: number } = {}
  ): Promise<Record<string, unknown>> {
    const url = `${this._baseUrl}${apiPath}`;
    const timeoutMs = (opts.timeout ?? 10) * 1000;

    let resp: Response;
    try {
      resp = await fetch(url, {
        method,
        headers: this._headers,
        body: body !== undefined ? JSON.stringify(body) : undefined,
        signal: AbortSignal.timeout(timeoutMs),
      });
    } catch (err) {
      if (err instanceof Error && err.name === "AbortError") {
        throw new LiveWorkbenchError(
          `Timeout ao chamar ${apiPath} (>${opts.timeout ?? 10}s)`
        );
      }
      throw new LiveWorkbenchError(
        `Servidor vivo indisponível em ${this._baseUrl}. Rode 'arkhe start' e tente novamente.`
      );
    }

    if (!resp.ok) {
      let detail = "";
      try {
        detail = await resp.text();
      } catch {
        // ignore
      }
      throw new LiveWorkbenchError(
        `HTTP ${resp.status} em ${apiPath}: ${detail.slice(0, 300)}`
      );
    }

    let payload: unknown;
    try {
      payload = await resp.json();
    } catch {
      throw new LiveWorkbenchError(`Resposta inválida (não-JSON) do backend vivo em ${apiPath}`);
    }

    if (typeof payload !== "object" || payload === null || Array.isArray(payload)) {
      throw new LiveWorkbenchError(`Resposta inválida do backend vivo em ${apiPath}`);
    }

    return payload as Record<string, unknown>;
  }
}
